File: src/main/java/easy/question2/ValidSubsequence.java
package easy.question2;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ValidSubsequence {
    public static void main(String[] args) {
        System.out.println(isValidSubsequence(
                new ArrayList<>(Arrays.asList(5, 1, 22, 25, 6, -1, 8, 10)),
                new ArrayList<>(Arrays.asList(1, 6, -1, 10))
        ));
    }

    public static boolean isValidSubsequence(List<Integer> array, List<Integer> sequence) {
        int arrIdx = 0;
        int seqIdx = 0;
        while(arrIdx < array.size() && seqIdx < sequence.size()){
            if(array.get(arrIdx).equals(sequence.get(seqIdx))){
                seqIdx++;
            }
            arrIdx++;
        }
        return seqIdx == sequence.size();
    }

    public static boolean isValidSubsequence1(List<Integer> array, List<Integer> sequence) {
        int seqIdx = 0;
        for(var value: array){
            if(seqIdx == sequence.size()){
                break;
            }
            if(sequence.get(seqIdx).equals(value)){
                seqIdx++;
            }
        }
        return seqIdx == sequence.size();
    }
}

File: src/main/java/easy/question2/ValidSubsequence.scala
package easy.question2

import scala.util.control.Breaks.break


object ValidSubsequenceS extends App {
  //ValidSubsequenceS().isValidSubsequence(List[Int](5, 1, 22, 25, 6, -1, 8, 10), List[Int](1, 6, -1, 10))

  println(ValidSubsequenceS().lcs(List[Int](5, 1, 22, 25, 6, -1, 8, 10), List[Int](1, 6, -1, 10)))
}

case class ValidSubsequenceS() {
  def isValidSubsequence(array: List[Int], sequence: List[Int]) = {
    var seqIdx = 0
    for(v <- array){
      if (seqIdx == sequence.length) break
      if(sequence(seqIdx) == v) seqIdx +=1
    }
    seqIdx == sequence.length
  }

  def lcs[T]: (List[T], List[T]) => List[T] = {
    case (_, Nil) => Nil
    case (Nil, _) => Nil
    case (x :: xs, y :: ys) if x == y => x :: lcs(xs, ys)
    case (x :: xs, y :: ys)           => {
      (lcs(x :: xs, ys), lcs(xs, y :: ys)) match {
        case (xs, ys) if xs.length > ys.length => xs
        case (xs, ys)                          => ys
      }
    }
  }
}
